### ISSUE TITLE
Tracking issue for io_slice_advance

### ISSUE THREAD

**Issue by Thomasdezeeuw** [rust-lang/rust#62726]
This is a tracking issue for `IoSlice::{advance, advance_slices}` and `IoSliceMut::{advance, advance_slices}`.

Feature gate: #![feature(io_slice_advance)].

Steps:

- [x] Implement the RFC.
    * #62987
    * #85802
- [ ] FCP
- [ ] Stabilization PR.

Current API additions:

```rust
impl<'a> IoSlice<'a> { // And `IoSliceMut`
    pub fn advance(&mut self, n: usize);
    pub fn advance_slices(bufs: &mut &mut [IoSlice<'a>], n: usize);
}
```

---
Old issue:

Writing https://github.com/rust-lang-nursery/futures-rs/pull/1741 I needed to resort to unsafe code to change the underlying slice in `IoSlice` (and `IoSliceMut`). I'm missing a method that can change the underlying slice. @Nemo157 said that I should open an issue.

Current idea would be something like `Buf::advance` from the [bytes crate](https://docs.rs/bytes/0.4.12/bytes/trait.Buf.html#tymethod.advance).

```rust
impl IoSlice {
    // Advance the internal cursors of the slice by `n` bytes.
    fn advance(&mut self, n: usize) {
        // ..
    }
}
```

**Comment by Thomasdezeeuw** [comment]
Looking for some feedback as to whether this would be accepted.

**Comment by Thomasdezeeuw** [comment]
Can someone reopen this as the tracking issue for `io_slice_advance`? I've updated the the first comment, but I'm not sure it's 100% correct.

**Comment by seanmonstar** [comment]
I was trying to `impl Buf for IoSlice` in the bytes crate, and noticed there's currently an issue with lifetimes:

```rust
impl Buf for IoSlice<'_> {
    fn advance(&mut self, cnt: usize) {
        *self = IoSlice::new(&self[cnt..]); // conflicting lifetime requirements
    }
}
```

To allow such a thing externally, we could add a method to `IoSlice` that gives the slice with the same lifetime as `IoSlice` (whereas `Deref` uses an anonymous lifetime).

```rust
impl<'a> IoSlice<'a> {
    pub fn inner_ref(&self) -> &'a [u8] {
        // ...
    }
}
```

**Comment by seanmonstar** [comment]
@Thomasdezeeuw I don't *think* so. It's not possible to get the inner slice with the same lifetime due to `Deref`:

```rust
let tmp = std::mem::replace(self, IoSlice::new(&[]));
let tail = &tmp[cnt..]; // <- lifetime isn't same as `IoSlice<'a>`
*self = IoSlice::new(tail);
```

**Comment by Nemo157** [comment]
@seanmonstar it's not an _amazing_ API to build that on top of, but it is possible ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=8bee4f76e80031f9100a5290cef5b606)):

```rust
let mut slices = [std::mem::replace(self, IoSlice::new(&[]))];
let slices = IoSlice::advance(&mut slices[..], cnt);
match slices {
    [slice] => {
        *self = std::mem::replace(slice, IoSlice::new(&[]));
    }
    [] => {
        *self = IoSlice::new(&[]);
    }
    _ => unreachable!(),
}
```

**Comment by seanmonstar** [comment]
Oh right, that is using the unstable function. I was referring to be able to implement this now.

**Comment by Nemo157** [comment]
Ah, I just assumed since that's what the tracking issue is about. It seems like having `IoSlice<'a>::into_inner(self) -> &'a [u8]` along with `Clone` and `Copy` would make sense since it's just a newtyped slice.

**Comment by noctune** [comment]
I've been trying to write a `write_all_vectored` function that continuously calls `write_vectored`, but it seems like this is not really possible without some way of advancing the `IoSlice`s. This feature would be very useful in making vectored IO more practical.

**Comment by Ekleog** [comment]
Are there any news on landing this? I'm trying to do vectored IO too, and it would be very nice if it were possible to just use this function — I think this is the only unstable feature I currently depend on, and it would be very nice to be able to compile on stable without having to add unsafe code.

`into_inner` as discussed above is another primitive that would make it possible to implement `advance` as a user, and it would most likely be great to have it, but seeing the way vectored IO is done I can't see a world where a function like `advance` isn't required.

**Comment by m-ou-se** [comment]
The signature of the unstable function in std right now is:
```rust
pub fn advance<'b>(
    bufs: &'b mut [IoSlice<'a>],
    n: usize
) -> &'b mut [IoSlice<'a>]
```

It might be a bit confusing that it *both* modifies `bufs` (somtimes at least), but also returns the new slice. A `#[must_use]` would be good.

An alternative could be
```rust
pub fn advance(
    bufs: &mut &mut [IoSlice<'a>],
    n: usize
)
```

**Comment by Thomasdezeeuw** [comment]
@m-ou-se I agree, I think previous versions didn't return a value. But I can't remember why it returns a value now. If other people agree I could send a pr.

**Comment by Ekleog** [comment]
@m-ou-se Does the removal of the I-nominated label mean that libs team discussed it and concluded that the signature should change for the alternative you wrote out in your post?

(Just coming back here as AFAIK this feature is now the only one blocking me from building on rustc stable, through futures-utils' write-all-vectored feature. I can try to see whether I can PR into futures-utils so that the feature can be implemented without this function, but if there's a known way towards stabilization there's little point in spending that many cycles for it, especially as it'll probably mean having to do unsafe to increase lifetimes)

**Comment by m-ou-se** [comment]
@Ekleog - I removed the label because there didn't seem to be anything to discuss in a meeting. But if you think there's anything that needs a decision from the libs team I can put it back.

(If you have an opinion on the signature of this function, I'd love to hear it in a comment here. :) )

**Comment by Ekleog** [comment]
I must say I don't really have any strong opinions on the signature of this function, but thank you for asking! Both ways sound reasonable enough to me, as it's expected to be a low-level building block; though I must say that the alternative you suggested does “feel” slightly better to me (but I don't know whether this feeling is worth anything, especially as since futures-rs got its `write-all-vectored` feature I no longer even call this function myself). FWIW, digging into github makes it look like the API changed from `&mut &mut` to `&mut -> &mut` here (https://github.com/rust-lang/rust/pull/62987#discussion_r308299291), I'm not totally sure why either, maybe there was off-band discussion.

Overall, my comment was mostly to figure out the current state of stabilization and what it's blocked on; I guess it means it's currently blocked on refining the signature, and then following the usual process of waiting for some time for the ecosystem to settle, and if things look all good stabilizing — though I'm not sure by which process “refining the signature” should be happening, but it's the first time I really look at the stabilization process (as it's the first time I both depend on a single nightly feature and actually care about building on stable, I guess), so I suppose it will eventually happen following the normal procedure :)

Anyway, thank you for your feedback!

**Comment by Thomasdezeeuw** [comment]
I agree with @Ekleog I don't really care about the signature either, I'm ok with both. The `&mut &mut` version is uncommon in Rust, but I think the `&mut -> &mut` is a little more awkward to use at times. For example `WriteAllVectored` from futures calls `mem::take` before each call to `advance` (https://github.com/rust-lang/futures-rs/blob/54f795f7f2b95cc925580d8d3537e01f09db3a7b/futures-util/src/io/write_all_vectored.rs#L37).

Either way I think we should pick a version, accept it's not perfect, and stabilise it.

**Comment by joshtriplett** [comment]
@Thomasdezeeuw To reduce confusion, could you please edit the first post of this tracking issue to give the *current* signatures of `IoSlice::advance` and `IoSliceMut::advance`, rather than referencing an old proposal that doesn't match the current API?

**Comment by Thomasdezeeuw** [comment]
@joshtriplett Done.

**Comment by mpdn** [comment]
We could sidestep the signature issue by just implementing a method on the `IoSlice` itself like `advance(&mut self, n: usize)`. That would at least make it possible for the user to create a method with either signature. Basically giving the user the need-to-have without the nice-to-have. Having a method for advancing on the `IoSlice` also seems like it would be a good addition in isolation.

**Comment by Thomasdezeeuw** [comment]
@mpdn I implemented that on the underlying sys types, e.g: https://github.com/rust-lang/rust/blob/0d97f7a96877a96015d70ece41ad08bb7af12377/library/std/src/sys/unix/io.rs#L23-L32

**Comment by mpdn** [comment]
Oh nice. I now see the "old issue" in the issue is actually what I described. I think exposing this would be a nice MVP. Biggest problem is that it collides with the existing name. I guess the existing one could be renamed to something like `advance_slices`. Thoughts?

**Comment by Thomasdezeeuw** [comment]
I was thinking about this some more and I think having the possibility to advance both a single `IoSlice` and a slice of `IoSlice`s is quite useful, I've needed both a couple of times myself. So, I want propose a slight change in design to make both possible:

```rust
impl<'a> IoSlice<'a> { // And `IoSliceMut`.
    /// Advance the `IoSlice` by `n` bytes.
    pub fn advance(&mut self, n: usize) {
        // ...
    }

    /// Advance `bufs` by `n` bytes.
    // NOTE: can't (currently) call this as `self: &mut &mut [IoSlice]`.
    pub fn advance_slice(bufs: &mut &mut [IoSlice<'a>], n: usize) {
        // ...
    }
}
```

**Comment by Thomasdezeeuw** [comment]
I've created a pr for the changes in https://github.com/rust-lang/rust/issues/62726#issuecomment-848610798: https://github.com/rust-lang/rust/pull/85802

**Comment by zethra** [comment]
I was wondering what the status of this features is? I could really use this API in something I'm working on.

**Comment by Thomasdezeeuw** [comment]
I've being using it without issues, so from my point of view this is ready for a FCP as the last activity was back in June (#85802).

**Comment by kalcutter** [comment]
Please FCP this. It is quite annoying implementing functions that do vectored I/O without any straightforward way to advance.

**Comment by m-ou-se** [comment]
Right now, `advance` panics if its argument is larger than the amount of bytes in the slice(s), but `advance_slices` does not. Both should probably have the same behaviour, and it should be documented.

Other than that, this seems ready for stabilization.

**Comment by m-ou-se** [comment]
I think both of them should panic in that case, but am open to hear why the non-panicking behaviour would make more sense.

**Comment by m-ou-se** [comment]
PR for that: https://github.com/rust-lang/rust/pull/94855

**Comment by kalcutter** [comment]
I think the panic behavior is reasonable since these functions are analogous to `slice = &slice[n..]`. On the other hand, there is also unstable `Iterattor::advance_by` which returns a `Result<(), usize>` instead. However, for an iterator it may not be possible to know the total number of items whereas for `IoSlice` the user may always check `len()`.

**Comment by kalcutter** [comment]
Are there any concerns with the current panic behavior? If not, can this be FCPed?

**Comment by PureWhiteWu** [comment]
Really looking forward for this being stabilized. This is quite useful in network programming.

**Comment by dead-claudia** [comment]
I do have one quick question prior to it stabilizing: do you want to require people to do `if bufs.is_empty()`/`if buf.is_empty()`, or should both methods return booleans, returning `true` if the buffer is non-empty and `false` otherwise? The latter would be make calling code a little easier to follow IMHO, since you can just use it as the `while` condition in most loops:

```rust
fn write_all_vectored(&mut self, mut bufs: &mut [IoSlice<'_>]) -> Result<()> {
    // Advance initially with `n = 0` to guarantee that bufs is empty if it
    // contains no data, to avoid calling write_vectored if there is no data
    // to be written.
    let mut n = 0;
    while IoSlice::advance_slices(&mut bufs, n) {
        match self.write_vectored(bufs) {
            Ok(0) => {
                return Err(error::const_io_error!(
                    ErrorKind::WriteZero,
                    "failed to write whole buffer",
                ));
            }
            Ok(bytes) => n = bytes,
            Err(ref e) if e.kind() == ErrorKind::Interrupted => {}
            Err(e) => return Err(e),
        }
    }
    Ok(())
}
```

The code itself [already necessarily checks for this for the outer buffer](https://github.com/rust-lang/rust/blob/master/library/std/src/io/mod.rs#L1174), and it's pretty trivial for the inner buffer.

**Comment by kalcutter** [comment]
How can progress be made on this issue? It is unclear to me what concerns are blocking this.

**Comment by Thomasdezeeuw** [comment]
@kalcutter  it hasn't seen any changes for a year and half (?), so I think someone just needs to create a stabilisation pr.

**Comment by eduardosm** [comment]
I think a FCP is needed first.

**Comment by Thomasdezeeuw** [comment]
> I think a FCP is needed first.

Isn't that usually done in the stabilization pr?

**Comment by eduardosm** [comment]
I'm not sure, but I have usually seen FCPs in tracking issues.

**Comment by andll** [comment]
+1 for starting FCP, this is a really useful feature

**Comment by Amanieu** [comment]
This is a generally useful API for working with vectored I/O.

@rfcbot fcp merge

---

> I do have one quick question prior to it stabilizing: do you want to require people to do `if bufs.is_empty()`/`if buf.is_empty()`, or should both methods return booleans, returning `true` if the buffer is non-empty and `false` otherwise? The latter would be make calling code a little easier to follow IMHO, since you can just use it as the `while` condition in most loops:

I think making `advance`/`advance_slices` return a bool would be confusing, it isn't immediately clear from the function name what the return value means. Code is much clearer if it explicitly queries `is_empty` on the slice instead.

**Comment by BurntSushi** [comment]
The `advance` method is an inherent method on a type that implements `Deref` for `[u8]`. Don't we usually try to avoid adding new inherent methods in that case? (I suppose this would only become an issue if we added a method named `advance` to `[u8]`. Seems unlikely but not completely impossible?)

**Comment by Mark-Simulacrum** [comment]
FWIW, bytes provides advance on &[u8] via https://docs.rs/bytes/latest/bytes/trait.Buf.html. Is there a reason not to put advance directly on &[u8] here? I think that would be slightly incompatible to move later...

**Comment by dead-claudia** [comment]
> FWIW, bytes provides advance on &[u8] via https://docs.rs/bytes/latest/bytes/trait.Buf.html. Is there a reason not to put advance directly on &[u8] here? I think that would be slightly incompatible to move later...

Arguably it's inconsistent, since the other method takes a slice of `IoSlice`s.

However, I could see some benefit of moving both to methods via a pair of extension traits (say, `ReadTarget` and `WriteTarget`):

- You could provide overloads for both `IoSlice`/`IoSliceMut` and `&[u8]`/`&mut [u8]`, all using the same name.
- These traits could also include `ReadTarget::read_from(&mut self, impl Read)` and `WriteTarget::write_into(&mut self, impl Write)` methods, and those could be used to be polymorphic over normal and vectorized reads, easing migration between the two a *lot*.

**Comment by eduardosm** [comment]
I opened https://github.com/rust-lang/rust/pull/116070 to handle that.

**Comment by thomaseizinger** [comment]
> > > FWIW, bytes provides advance on &[u8] via [docs.rs/bytes/latest/bytes/trait.Buf.html](https://docs.rs/bytes/latest/bytes/trait.Buf.html). Is there a reason not to put advance directly on &[u8] here? I think that would be slightly incompatible to move later...
> > 
> > 
> > Arguably it's inconsistent, since the other method takes a slice of `IoSlice`s.
> 
> That's because the `advance_slices` method works on a slice of `IoSlice` (hence the name), thus needs mutable access to `&mut [IoSlice]`, which means `&mut &mut [IoSlice]`, which doesn't work with a variant of `&mut self`.
> 
> That said, I'm perfectly fine with changing `fn advance(&mut self, n: usize)` to `fn advance(this: &mut IoSlice<'a>, n: usize)` to avoid this issue mentioned by @BurntSushi.

Although it is also still unstable, a related set of functions on `MaybeUninit` uses a different naming pattern: https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#method.slice_as_bytes

```rust
impl<T> MaybeUninit<T> {
    pub fn slice_as_bytes(this: &[MaybeUninit<T>]) -> &[MaybeUninit<u8>];
}
```

Applying the same here would read kind of odd though:

```rust
impl<T> IoSlice<T> {
	pub fn slice_advance(this &mut &mut [IoSlice<'a>], n: usize);
}

// in some impl
IoSlice::slice_advance(&mut bufs, n_read);
```

Perhaps we could name them `advance_one` and `advance_many`?

**Comment by Thomasdezeeuw** [comment]
> Although it is also still unstable, a related set of functions on `MaybeUninit` uses a different naming pattern: https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#method.slice_as_bytes
> 
> ```rust
> impl<T> MaybeUninit<T> {
>     pub fn slice_as_bytes(this: &[MaybeUninit<T>]) -> &[MaybeUninit<u8>];
> }
> ```

I fail to see the related between the two functions. `MaybeUninit:: slice_as_bytes` does a (type) conversion, while `IoSlice::slice_advance` manipulates the slices, I don't think they should follow the same naming pattern.

Can you elaborate why you think they should?

**Comment by thomaseizinger** [comment]
> I fail to see the relation between the two functions. `MaybeUninit:: slice_as_bytes` does a (type) conversion, while `IoSlice::slice_advance` manipulates the slices, I don't think they should follow the same naming pattern.
> 
> Can you elaborate why you think they should?

`as_bytes` is the "action" of the function and equivalent of "advance" for `IoSlice`. It is prefixed by `slice_` because the first argument isn't `self` but a slice of `self`. Following that logic, the functions should be:

```rust
impl IoSlice {
	fn advance(&mut self, n: usize);
	fn slice_advance(this &mut &mut [IoSlice<'a>], n: usize);
}
```

Perhaps it should be a free-function instead and we offer only `IoSlice::advance`?

**Comment by thomaseizinger** [comment]
It is also worth noting that `AsyncWrite::poll_write_vectored` and `Write::write_vectored` only give you a `&[IoSlice<'_>]`. How are implementations meant to advance these anyway?

**Comment by Thomasdezeeuw** [comment]
> > I fail to see the relation between the two functions. `MaybeUninit:: slice_as_bytes` does a (type) conversion, while `IoSlice::slice_advance` manipulates the slices, I don't think they should follow the same naming pattern.
> > Can you elaborate why you think they should?
> 
> `as_bytes` is the "action" of the function and equivalent of "advance" for `IoSlice`. It is prefixed by `slice_` because the first argument isn't `self` but a slice of `self`. Following that logic, the functions should be:
> 
> ```rust
> impl IoSlice {
> 	fn advance(&mut self, n: usize);
> 	fn slice_advance(this &mut &mut [IoSlice<'a>], n: usize);
> }
> ```

I still fail to see the relation, but those names seem fine with me as well. At this point I don't really care about the names to be honest.

> It is also worth noting that `AsyncWrite::poll_write_vectored` and `Write::write_vectored` only give you a `&[IoSlice<'_>]`. How are implementations meant to advance these anyway?

See [`io::Write::vectored_all`](https://doc.rust-lang.org/std/io/trait.Write.html#method.write_all_vectored), `write_vectored` shouldn't advance the buffers.

**Comment by c410-f3r** [comment]
What is the status of this feature? There are no concerns in the FCP.

**Comment by PureWhiteWu** [comment]
Hi, is there any progress of this feature?

**Comment by zh-jq-b** [comment]
The following assert won't work after this if the `reader` call IoSliceMut::advance() in it's `read_vectored()`.
```rust
fn what_to_read<R>(reader: R)
where
    R: Read,
{
    let mut b1 = [0u8; 16];
    let mut b2 = [0u8; 32];

    let mut iov = [IoSliceMut::new(&mut b1), IoSliceMut::new(&mut b2)];
    let nr = reader.read_vectored(&mut iov).unwrap();

    if nr > 0 {
        assert_eq!(b1[0], iov[0][0]);
    }
    if nr >= b1.len() {
        assert_eq!(iov[0].len(), b1.len());
    }
}
```

**Comment by Thomasdezeeuw** [comment]
@zh-jq-b in your example you don't call `IoSliceMut::advance`, so I don't see how it's relevant to the method?

**Comment by zh-jq-b** [comment]
> @zh-jq-b in your example you don't call `IoSliceMut::advance`, so I don't see how it's relevant to the method?

Yes I didn't call `IoSliceMut::advance` but the `reader.read_vectored()` may call that indirectly?
What I want to say is that, there is no guranteen that `IoSliceMut::advance` won't get called inside `reader.read_vectored()`.

**Comment by Thomasdezeeuw** [comment]
I don't think we can guarantee that the method will never be incorrectly used. Same for `process::exit` for example... I don't know what you expect the standard library to change about this.

**Comment by zh-jq-b** [comment]
> I don't think we can guarantee that the method will never be incorrectly used. Same for `process::exit` for example... I don't know what you expect the standard library to change about this.

Before the stabilization of io_slice_advance, users can use both the original buffer and also the `iov` to get the read data.
After the stabilization, I'm confused with which one should be used?
For all old methods that take a `IoSliceMut`, should I still read the data by using the old way? Will it broken without noticeable API changes?

**Comment by dead-claudia** [comment]
@zh-jq-b Consider filing a bug report against anyone who sets them incorrectly.

It's not Rust's fault if someone implements `read_vectored` wrong. It's the fault of whoever wrote the bad implementation for that `io::Read` trait instance method.

**Comment by zh-jq-b** [comment]
> @zh-jq-b Consider filing a bug report against anyone who sets them incorrectly.
> 
> It's not Rust's fault if someone implements `read_vectored` wrong. It's the fault of whoever wrote the bad implementation for that `io::Read` trait instance method.

The `read_vectored` should be documented to specify what's the wrong implementation. So I opened #128669 for further confirmation.
